### PR TITLE
z3: Update to 4.12.1

### PR DESCRIPTION
**Pull request by eschnett**

### CODE REVIEW

**Comment by eschnett** [comment]
This patch changes the way `GMP` is installed: it removes the call to `get_addable_spec`. Without this change, no matching `GMP_jll` would be found, and GMP would not be installed, making the build fail.

File: Z/z3/build_tarballs.jl [hunk 1]
```diff
@@ -69,7 +70,7 @@ products = [
 # Dependencies that must be installed before this package can be built
 dependencies = [
     BuildDependency("libjulia_jll"),
-    Dependency(get_addable_spec("GMP_jll", v"6.2.0+5")),
+    Dependency("GMP_jll"),
```
Comment (giordano) [inline Z/z3/build_tarballs.jl:73]: We still have to compile with an old GMP, because of compatibility reasons.

Comment (eschnett) [reply Z/z3/build_tarballs.jl:73]: Which GMP should I use? The previous `6.2.0+5` did not work.

Comment (giordano) [reply Z/z3/build_tarballs.jl:73]: I'd need to understand _why_ it didn't work.

Comment (eschnett) [reply Z/z3/build_tarballs.jl:73]: The log message was
```
┌ Warning: Dependency GMP_jll does not have an (Stdlib)Artifacts.toml in /Users/eschnett/.julia/packages/GMP_jll/IA6Pi!
└ @ BinaryBuilderBase ~/.julia/dev/BinaryBuilderBase/src/Prefix.jl:742
```

Afterwards, no GMP was installed, and cmake failed.

Comment (eschnett) [reply Z/z3/build_tarballs.jl:73]: I used Julia 1.8.5 and the `master` versions of `BinaryBuilder` and `BinaryBuilderBase`.

Comment (giordano) [inline Z/z3/build_tarballs.jl:73]: Let's try with
```suggestion
    Dependency("GMP_jll"; compat="6.2.0"),
```

**Pull request merged**